### DISCUSSION TITLE
html minify plugin

### DISCUSSION THREAD

**Discussion by RWDevelopment** [BoidCMS/BoidCMS#21] [Q&A] [answered]
any hints for making html minify plugin or function for theme ?
how to do it ?

i'm making new admin dashboard, plugins and themes for BoidCMS & trying to learn this great cms
![screencapture-boidcms-admin-2022-09-15-22_16_51](https://user-images.githubusercontent.com/33756545/190500570-eb8c866f-4d31-45d8-a3fb-c2e2f9143700.jpg)

**Comment by sysadev** [comment, marked as answer]
Hi @RWDevelopment, sorry for the late response.

BoidCMS doesn't have any built-in method for accessing or modifying the HTML output, using the PHP output buffering `ob_start` and the `render` and `rendered` actions will work in this kinda situation.

- The `render` action is the first triggered action on every request.
- The `rendered` action is the last fired action.

Sample:
```php
<?php

global $App;
$App->set_action( 'render', 'my_minify_ob_start' );
$App->set_action( 'rendered', 'my_minify_ob_end' );

function my_minify_ob_start() {
  // @see https://www.php.net/manual/en/function.ob-gzhandler.php
  ob_start( 'ob_gzhandler' );
}

function my_minify_ob_end() {
  ob_end_flush();
}

?>
```

Thanks for the interest in contributing to this awesome cms. And I'll be happy to help you out.

**Comment by RWDevelopment** [comment]
Thanks for response, below is my first attempt to html minify plugin:

```
<?php

global $App;
$App->set_action( 'render', 'my_minify_ob_start' );
$App->set_action( 'rendered', 'my_minify_ob_end' );

function my_minify_ob_start() {
    ob_start('minifier');
}

function my_minify_ob_end() {
    ob_end_flush();
}

function minifier($code) {
  $search = array(
       
      // Remove whitespaces after tags
      '/\>[^\S ]+/s',
       
      // Remove whitespaces before tags
      '/[^\S ]+\</s',
       
      // Remove multiple whitespace sequences
      '/(\s)+/s',
       
      // Removes comments
      '/<!--(.|\s)*?-->/'
  );
  $replace = array('>', '<', '\\1');
  $code = preg_replace($search, $replace, $code);
  return $code;
}

?>
```

**Reply by sysadev** [reply]
Awesome
